test/model/tools/model_test/main.cpp

```diff
@@ -244,6 +249,146 @@ load_workload(const char *file)
+                if (itr == txn_to_sequence.end())
+                    throw model::model_exception(
+                      "Transaction ID does not exist: " + std::to_string(txn_id));
+                model::kv_workload_sequence_ptr seq = itr->second;
+                *seq.get() << op.operation;
+                op.seq_no = seq->seq_no();
+
+                /* Transaction end. */
+                if (std::holds_alternative<model::operation::commit_transaction>(op.operation) ||
+                  std::holds_alternative<model::operation::rollback_transaction>(op.operation))
+                    txn_to_sequence.erase(itr);
+            }
+        } else {
+            /* Non-transaction operations, such as set stable timestamp or crash. */
+            model::kv_workload_sequence_ptr seq =
+              std::make_shared<model::kv_workload_sequence>(sequences.size());
+            *seq.get() << op.operation;
+            op.seq_no = seq->seq_no();
+            sequences.push_back(seq);
+
+            /* Operations that clear the transaction state. */
+            if (std::holds_alternative<model::operation::crash>(op.operation) ||
+              std::holds_alternative<model::operation::restart>(op.operation))
+                txn_to_sequence.clear();
+        }
+    }
+
+    /*
+     * Find the minimal subset that causes a failure by using an algorithm similar too binary
+     * search: Remove the first half of the workload and then see if the failure happens again. If
+     * it does, then that part of the workload can be eliminated. If the failure does not happen, it
+     * means that the eliminated part of the workload includes something that contributes to the
+     * failure. In the next iteration, try removing only half of the removed range - first the first
+     * half and then the second half.
+     */
+    std::vector<bool> enabled(sequences.size(), true);
+    std::shared_ptr<model::kv_workload> reduced;
+
+    std::deque<std::pair<size_t, size_t>> ranges_to_remove;
+    ranges_to_remove.push_back(std::pair<size_t, size_t>(0, sequences.size() / 2));
+    ranges_to_remove.push_back(std::pair<size_t, size_t>(sequences.size() / 2, sequences.size()));
+
+    size_t round = 0;
+    while (!ranges_to_remove.empty()) {
+        std::pair<size_t, size_t> range = ranges_to_remove.front();
+        ranges_to_remove.pop_front();
+        if (range.first >= range.second)
+            continue;
+
+        round++;
+        std::cout << "Counterexample reduction: Round " << round << ", remove " << range.first
+                  << "-" << range.second << std::endl;
+
+        /* Create a workload with a subset of the operations. */
+        std::shared_ptr<model::kv_workload> w = std::make_shared<model::kv_workload>();
+        for (size_t i = 0; i < workload->size(); i++) {
+            const model::kv_workload_operation &op = (*workload)[i];
+
+            if (op.seq_no == model::k_no_seq_no ||
```

Comment:
Perhaps a comment for the `k_no_seq_no` check - I found it illuminating when you explained this was only here as a precaution.

Reply:
Fixed in [99d0adb](https://github.com/wiredtiger/wiredtiger/pull/10318/commits/99d0adb57880cb11dd452ea30ba81ee3649e1e9b).